qutip/core/blochredfield.py

```diff
@@ -0,0 +1,162 @@
+import os
+import numpy as np
+from qutip.settings import settings as qset
+
+from . import Qobj, QobjEvo, liouvillian, Cubic_Spline, coefficient, sprepost
+from ._brtools import SpectraCoefficient, _EigenBasisTransform
+from ._brtensor import _BlochRedfieldElement
+
+
+__all__ = ['bloch_redfield_tensor', 'brterm']
+
+
+def bloch_redfield_tensor(H, a_ops, c_ops=[], sec_cutoff=0.1,
+                          fock_basis=False, sparse_eigensolver=False,
+                          br_dtype='sparse'):
+    """
+    Calculates the Bloch-Redfield tensor for a system given
+    a set of operators and corresponding spectral functions that describes the
+    system's coupling to its environment.
+
+    Parameters
+    ----------
+
+    H : :class:`qutip.Qobj`, :class:`qutip.QobjEvo`
+        System Hamiltonian.
+
+    a_ops : list of (a_op, spectra)
+        Nested list of system operators that couple to the environment,
+        and the corresponding bath spectra.
+
+        a_op : :class:`qutip.Qobj`, :class:`qutip.QobjEvo`
+            The operator coupling to the environment. Must be hermitian.
+
+        spectra : :class:`Coefficient`
+            The corresponding bath spectra.
+            Can be a `Coefficient` using an 'w' args or a function of the
+            frequency. The `SpectraCoefficient` can be used to use array based
+            coefficient. They can also depend on ``t`` if the corresponding
+            ``a_op`` is a :cls:`QobjEvo`.
+
+        Example:
+            a_ops = [
+                (a+a.dag(), coefficient('w>0', args={"w": 0})),
+                (QobjEvo([b+b.dag(), f(t)]), g(w)),
+                (c+c.dag(), SpectraCoefficient(coefficient(Cubic_Spline))),
```

Comment:
I think the new structure of `a_ops` is nice, but I think the documentation here is currently a little imprecise.

- the `a_op` always seems to be converted to a QobjEvo and the spectra is always called with `spectra(t, w=...)` so I'm not sure that functions that depend only on `w` will work?
- the example passes `f(t)` and `g(w)` but presumably one would type `[..., f]` and `(QobjEvo(...), g)` and the `t` and `w` are meant to indicate the signatures of the two functions. Perhaps the signatures could be indicated another way?
- the long `SpectraCoefficient(coefficient(Cubic_Spline))` is a bit unwieldy and I'm not sure what `Cubic_Spline` is supposed to be (maybe this was written before the new order parameter for array  coefficients)

It would be good to have a nicer way to pass in an ndarray of coefficients and perhaps specify the order.

Would it be worthwhile trying to support a 2-d ndarray for the time-dependent case (so that would could specify a mesh of values)? Perhaps that is just too messy -- mesh interpolation is much more involved that 1-d interpolation.

The example implies that the code could be written roughly as is, but that is not the case for the last two elements of a_ops.

Reply:
Functions of `w` only are accepted in `brmesolve` and only there. Here it need to be already a `Coefficient`, good catch.

The way I see to have the signature clear is to use lambda, I changed it, tell me if you have a better idea.

`Cubic_Spline` is sometime we are removing with #1725. I merged `dev,major` and remove all it's use.

For `ndarray`, the user need to have created the `Coefficient` before calling this, so he can set the order then. As for 2-d ndarrays, I don't think it's worth it. As you say, it a lot more messy that 1-d and it's use case is limited, so for those cases using a function or string is enough. If there is demand for it we can add it later.